sslyze/plugins/openssl_cipher_suites_plugin.py

```diff
@@ -280,55 +289,113 @@ def _test_cipher_suite(
         return cipher_result
 
     @classmethod
-    def _get_preferred_cipher_suite(
+    def _get_preferred_cipher_suite_order(
             cls,
             server_connectivity_info: ServerConnectivityInfo,
             ssl_version: OpenSslVersionEnum,
             accepted_cipher_list: List['AcceptedCipherSuite']
-    ) -> Optional['AcceptedCipherSuite']:
-        """Try to detect the server's preferred cipher suite among all cipher suites supported by SSLyze.
+    ) -> Optional[List['AcceptedCipherSuite']]:
+        """Try to detect the server's preferred cipher suite order among all cipher suites supported by SSLyze.
+
+        The algorithm for determining the server's full preferred cipher suite
+        order is as follows:
+
+        1. Send the unordered list to the server.
+        2. Move the accepted cipher to the ordered list.
+        3. If the accepted cipher was the first one in the list, add it to
+           the end of the cipher string and send again to verify server
+           ordering is enabled.
+        4. If the server did not give us the same cipher it chose last time:
+               the server has no preference; stop and return `None`
+           Else:
+               remove the previously accepted cipher from the unordered list and continue
+        5. Send the unordered list again, moving the accepted cipher from the
+           unordered list to the ordered one. Repeat until the unordered list is empty.
         """
         if len(accepted_cipher_list) < 2:
             return None
 
-        accepted_cipher_names = [cipher.openssl_name for cipher in accepted_cipher_list]
+        unordered_cipher_names = [cipher.openssl_name for cipher in accepted_cipher_list]
+        server_ordered_cipher_list = []
         should_use_legacy_openssl = None
 
         # For TLS 1.2, we need to figure whether the modern or legacy OpenSSL should be used to connect
         if ssl_version == OpenSslVersionEnum.TLSV1_2:
             should_use_legacy_openssl = True
-            # If there are more than two modern-supported cipher suites, use the modern OpenSSL
-            for cipher_name in accepted_cipher_names:
-                modern_supported_cipher_count = 0
```

Comment:
I believe there was a bug (?) on line 302, where the `modern_supported_cipher_count` was reset back to 0 with every iteration, so that the modern client would never be used. The new code does invoke the modern client as needed, so there are some additional changes below around handling new corner cases.